// pw_server/pw_gameserver.js
/*

	pw_gameserver.js

*/

GameServer = (function (server) {

	this.server = server;

	this.activeGames = {};

	this.createGame = function (player1, player2) {

		var gid = this.generateGameId();
		var game = new Game (gid, this, player1, player2);
		this.activeGames[gid] = (game);
		return gid;

	};

	this.generateGameId = function () {

		var id = new String(Math.random()).slice(2);
		if (this.activeGames[id] === undefined) {
			return id;
		} else {
			return this.generateGameId();
		}

	};

	this.handleGameInfo = function (data) {

		if (data.gid === undefined || this.activeGames[data.gid] === undefined) {
			return;
		}

		if (data.action === GAME.ACTION.PRIVATE_CHAT) {
			var gid = data.gid;
			var pid1 = this.activeGames[gid].players[0].getID();
			var pid2 = this.activeGames[gid].players[1].getID();
			server.emitGameInfo(pid1, data);
			server.emitGameInfo(pid2, data);
		}

	};

	this.emitGameInfo = function (gid, data) {

	}

});

Game = (function (gid, gameServer, player1, player2) {

	this.gid = gid;
	this.gameServer = gameServer;
	this.players = new Array(2);
	this.players[0] = player1;
	this.players[1] = player2;

	this.emitPrivateMessage = function (data) {
		players[0].getSocket().emit({

		})
	};

});

module.exports = GameServer;
