// pw_server/pw_server.js
/*

	pw_server.js

*/

var util = require("util");
require("./pw_client");
require("../../../client/js/events");
require("./pw_gameserver");

/*
	PW_Server
*/
var PW_Server = function () {

	var io = null;
	var clients = {};
	var gameServer = new GameServer ();

	var setIO = function (IO) { io = IO; };

	var userConnectHandler = function (data, socket) {

		if (typeof data === undefined)
			return;

		data.socket = socket;
		client = new PW_Client(data);
		clients[client.getID()] = client;
		socket.set("clientID", client.getID());
		socket.broadcast.emit(SERVER.USER_CONNECTED, {
			name : client.getName(),
			id : client.getID()
		});
		userListHandler({}, socket);

	};

	var userDisconnectHandler = function (socket) {

		socket.get("clientID", function(err, id) {
			client = clients[id];
			if (client === "undefined") { return; }
			delete clients[id];
			socket.broadcast.emit(SERVER.USER_DISCONNECTED, {
				name : client.getName(),
				id : client.getID()
			});
		});

	};

	var publicMessageHandler = function (data, socket) {

		if (typeof data === "undefined")
			return;
		socket.broadcast.emit(CHAT.PUBLIC_MESSAGE, data);

	};

	var userListHandler = function (data, socket) {

		var list = [];
		for (var c in clients) {
			client = clients[c];
			list.push({ name : client.getName(), id : client.getID() });
		}
		socket.emit(CHAT.USER_LIST, list);

	};

	var createGameHandler = function (data, socket) {

		socket.broadcast.emit(GAME.CREATE, data);

	};

	var deleteGameHandler = function (data, socket) {

		socket.broadcast.emit(GAME.DELETE, data);

	};

	var joinGameHandler = function (data, socket) {

		var creator = clients[data.opponentID];
		var creatorSocket = creator.getSocket();
		// notify that creator of the game is not looking for opponent any more
		deleteGameHandler({ name : data.opponentName, id : data.opponentID }, creatorSocket);
		var joiner = clients[data.joinerID];
		var joinerSocket = joiner.getSocket();
		var gameID = gameServer.createGame(creator, joiner);

		creatorSocket.emit(GAME.START, {
			gid : gameID,
			opponentID : joiner.getID(),
			opponentName : joiner.getName()
		});

		joinerSocket.emit(GAME.START, {
			gid : gameID,
			opponentID : creator.getID(),
			opponentName : joiner.getName()
		});

	};

	var bindListeners = function () {

		io.sockets.on("connection", function(socket) {

			socket.on(CLIENT.AUTH, function (data) {
				util.log(CLIENT.AUTH + " " + JSON.stringify(data));
				userConnectHandler(data, socket);
			});

			socket.on(CHAT.PUBLIC_MESSAGE, function (data) {
				util.log(CHAT.PUBLIC_MESSAGE + " " + JSON.stringify(data));
				publicMessageHandler(data, socket);
			});

			socket.on(CHAT.USER_LIST, function (data) {
				util.log(CHAT.USER_LIST + " " + JSON.stringify(data));
				userListHandler(data, socket);
			});

			socket.on(GAME.CREATE, function (data) {
				util.log(GAME.CREATE + " " + JSON.stringify(data));
				createGameHandler(data, socket);
			});

			socket.on(GAME.DELETE, function (data) {
				util.log(GAME.DELETE + " " + JSON.stringify(data));
				deleteGameHandler(data, socket);
			});

			socket.on(GAME.JOIN_REQUEST, function (data) {
				util.log(GAME.JOIN_REQUEST + " " + JSON.stringify(data));
				joinGameHandler(data, socket);
			});

			socket.on(GAME.INFO, function (data) {
				util.log(GAME.INFO + " " + JSON.stringify(data));
				gameServer.handleGameInfo(data);
			})

			socket.on("disconnect", function () {
				util.log("DISCONNECT");
				userDisconnectHandler(socket);
			});

		});

	};

	return {
		setIO : setIO,
		bindListeners : bindListeners
	};

};

module.exports = new PW_Server();
